johnzon-core/src/main/java/org/apache/johnzon/core/Buffered.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.johnzon.core;
+
+/**
+ * A <tt>Buffered</tt> is a source or destination of data that is buffered
+ * before writing or reading.  The bufferSize method allows all participants
+ * in the underlying stream to align on this buffer size for optimization.
+ *
+ * This interface is designed in the spirit of {@code java.io.Flushable} and
+ * {@code java.io.Closeable}
+ *
+ * @since 1.2.17
+ */
+public interface Buffered {
```

Comment:
My first instinct was Bufferable as well.  What swayed me away is the "able" interfaces are all describing an action someone can take on the object and if they don't take that action, the thing described is not done.  If you don't serialize a `Serializable` it isn't serialized.  If you don't close a `Closeable` it isn't closed.  `Flushable` objects of course do flush eventually, but the method it has still triggers an action.  So calling it `Bufferable` felt a bit off as there's no action for the user to do.  It's buffered whether or not the user does anything and calling `bufferSize` does nothing.  `Buffered` is a statement of fact and `bufferSize` returns that fact.

All that said, I'm not super married to it.  That's just the logic that got me there.

Reply:
Hmm, guess the same reasoning got me asking the question, while not handled by johnzon it is not buffered.
What if we quit that area and call it `StreamDescriptor` which would be more OO (i am an instance of stream descriptor so I can get meta about the stream/writer).

Reply:
I personally like `Buffered` as it is telling Johnzon a fact, "I am buffered".  There is a ByteArrayOutputStream inside the supplied SnippetWriter that Johnzon doesn't control.  Johnzon can look at the writers and outputstreams it is given and use the interface to ask, "are you buffered?" and if they are Johnzon can optimize its own buffering to match their buffer size.  So to me it makes sense as there's a conversation there: 

johnzon: are you buffered? 
writer: yes, I'm buffered.
johnzon:  Ok, what's your buffer size?
writer:  It's 50.
johnzon: great, I'll write to you 50 bytes at a time

That said, if you'd like to rename it after the merge I'm ok with that.

Reply:
Oh I see where you coming from. I was more seeing it as:

johnzon: do you want to be buffered?
writer: yes, I need a custom buffer.
johnzon: Ok, what's your buffer requested size?
writer: It's 50.
johnzon: great, I'll write to you 50 bytes at a time


let me try to PR on top of your PR but this is not the critical point, was more curious ;)

Reply:
Cool.  As I said to me it makes sense.  We could easily get a Writer that writes to a file and has a buffer in it optimized for the page size of disk.  This interface would be a great way for them to tell that fact to Johnzon so it can participate in the optimization.  Ideally, whatever interface name use use it is strongly tied to buffering and not generic and therefore has many other purposes in the future.  If we did that, then it would get complicated for someone to say, "Yes, I am buffered, but I am not those other things."